code/datums/components/trapdoor.dm

```diff
@@ -0,0 +1,171 @@
+
+///makes this file more legible
+#define IS_OPEN(parent) isgroundlessturf(parent)
+
+GLOBAL_LIST_EMPTY(unlinked_trapdoors)
+
+/**
+ * # trapdoor component!
+ *
+ * component attached to floors to turn them into trapdoors, a constructable trap that when signalled drops people to the level below.
+ * assembly code at the bottom of this file
+ */
+/datum/component/trapdoor
+	///assembly tied to this trapdoor
+	var/obj/item/assembly/trapdoor/assembly
+	///path of the turf this should change into when the assembly is pulsed. needed for openspace trapdoors knowing what to turn back into
+	var/trapdoor_turf_path = /turf/open/openspace
+
+/datum/component/trapdoor/Initialize(starts_open, trapdoor_turf_path, assembly)
+	if(!isopenturf(parent))
+		return COMPONENT_INCOMPATIBLE
+	trapdoor_turf_path = parent.type
+	if(starts_open)
+		try_opening()
+	src.assembly = assembly
+	if(!src.assembly)
+		GLOB.unlinked_trapdoors += src
+	src.trapdoor_turf_path = trapdoor_turf_path
+
+/datum/component/trapdoor/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_TURF_CHANGE, .proc/turf_changed_pre)
+
+/datum/component/trapdoor/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(assembly, COMSIG_ASSEMBLY_PULSED)
+	UnregisterSignal(parent, COMSIG_TURF_CHANGE)
+
+///called by linking remotes to tie an assembly to the trapdoor
+/datum/component/trapdoor/proc/link_with_assembly(obj/item/assembly/trapdoor/assembly)
+	src.assembly = assembly
+	GLOB.unlinked_trapdoors -= src
+	assembly.linked = TRUE
+	RegisterSignal(assembly, COMSIG_ASSEMBLY_PULSED, .proc/toggle_trapdoor)
+
+///signal called by our assembly being pulsed
+/datum/component/trapdoor/proc/toggle_trapdoor(datum/source)
+	SIGNAL_HANDLER
+	if(!IS_OPEN(parent))
+		try_opening()
+	else
+		try_closing()
+
+///signal called by turf changing
+/datum/component/trapdoor/proc/turf_changed_pre(datum/source, path, new_baseturfs, flags, post_change_callbacks)
+	SIGNAL_HANDLER
+	var/turf/open/dying_trapdoor = parent
+	if(!IS_OPEN(dying_trapdoor) && !IS_OPEN(path)) //not a process of the trapdoor, so this trapdoor has been destroyed
```

Comment:
probably needs to have one of the trapdoor mechanism breaking cases be if the new path is plating